File: apps/web/src/lib/api/verification.ts
import { ApiError, apiFetch } from "@/lib/api/client";

export interface ReadinessCheck {
  check_id: string;
  passed: boolean;
  reason: string;
  evidence?: Record<string, unknown> | null;
}

export interface ReadinessReport {
  passed: boolean;
  checks: ReadinessCheck[];
}

export interface FillDiffItem {
  field_key: string;
  label: string | null;
  proposed_redacted?: string | null;
  actual_redacted?: string | null;
  matched?: boolean | null;
  locator_strategy?: string | null;
}

export interface ReviewPackage {
  run_id: string;
  job_target_id: string;
  company: string;
  role: string;
  status: string;
  human_summary: string;
  readiness: ReadinessReport;
  fill_diffs: FillDiffItem[];
  screenshot_object_key?: string | null;
  resume_filename?: string | null;
  cover_letter_preview?: string | null;
  cover_letter?: {
    id: string;
    text: string;
    ats_score: number | null;
    keyword_coverage?: {
      present?: string[];
      missing?: string[];
      template_style?: string;
      voice_preset?: string;
    };
    template_style?: string;
    voice_preset?: string;
    locked_paragraphs?: number[];
    pdf_download_path?: string;
  } | null;
  checkpoint_id?: string | null;
  policy: string;
}

export interface SubmitResult {
  run_id: string;
  outcome: string;
  confirmation_text?: string | null;
  final_url?: string | null;
  note?: string | null;
  job_target_status?: string | null;
}

export async function fetchReviewPackage(jobTargetId: string): Promise<ReviewPackage | null> {
  try {
    return await apiFetch<ReviewPackage>(`/api/job-targets/${jobTargetId}/review`);
  } catch (err: unknown) {
    if (err instanceof ApiError && err.status === 404) {
      return null;
    }
    throw err;
  }
}

export async function fetchReviewPackageByRun(runId: string): Promise<ReviewPackage | null> {
  try {
    return await apiFetch<ReviewPackage>(`/api/application-runs/${runId}/review`);
  } catch (err: unknown) {
    if (err instanceof ApiError && err.status === 404) {
      return null;
    }
    throw err;
  }
}

export async function submitApplicationRun(
  runId: string,
  fixtureHtml?: string,
): Promise<SubmitResult> {
  return apiFetch<SubmitResult>(`/api/application-runs/${runId}/submit`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(fixtureHtml ? { fixture_html: fixtureHtml } : {}),
  });
}

export async function skipApplicationSubmit(runId: string): Promise<{ run_id: string; status: string }> {
  return apiFetch(`/api/application-runs/${runId}/skip-submit`, { method: "POST" });
}

File: packages/schemas/src/jober_schemas/verification.py
from typing import Any
from uuid import UUID

from jober_schemas.common import SchemaBase
from jober_schemas.enums import RunStatus


class ReadinessCheckRead(SchemaBase):
    check_id: str
    passed: bool
    reason: str
    evidence: dict[str, Any] | None = None


class ReadinessReportRead(SchemaBase):
    passed: bool
    checks: list[ReadinessCheckRead]


class VerifyReadyRead(SchemaBase):
    run_id: UUID
    status: RunStatus
    readiness: ReadinessReportRead
    human_summary: str | None = None
    gate: str | None = None


class FillDiffItemRead(SchemaBase):
    field_key: str
    label: str | None
    proposed_redacted: str | None = None
    actual_redacted: str | None = None
    matched: bool | None = None
    locator_strategy: str | None = None


class ReviewCoverLetterRead(SchemaBase):
    id: UUID
    text: str
    ats_score: float | None = None
    keyword_coverage: dict[str, Any] | None = None
    template_style: str | None = None
    voice_preset: str | None = None
    locked_paragraphs: list[int] = []
    pdf_download_path: str | None = None


class ReviewPackageRead(SchemaBase):
    run_id: UUID
    job_target_id: UUID
    company: str
    role: str
    status: RunStatus
    human_summary: str
    readiness: ReadinessReportRead
    fill_diffs: list[FillDiffItemRead]
    screenshot_object_key: str | None = None
    resume_filename: str | None = None
    cover_letter_preview: str | None = None
    cover_letter: ReviewCoverLetterRead | None = None
    checkpoint_id: UUID | None = None
    policy: str


class SubmitResultRead(SchemaBase):
    run_id: UUID
    outcome: str
    confirmation_text: str | None = None
    final_url: str | None = None
    note: str | None = None
    job_target_status: str | None = None
